### DISCUSSION TITLE
Adding full HW offload support

### DISCUSSION THREAD

**Discussion by ferasbi** [strongswan/strongswan#1194] [Ideas]
Hi,

(This is related to https://github.com/strongswan/strongswan/pull/188)

Until now the options available to user for HW offload were:

- hw_offload = no
- hw_offload = yes
- hw_offload = auto

We would like to suggest adding a new HW offload option:
hw_offload = full
When choosing this option, we add to the offload flags the flag XFRM_OFFLOAD_FULL.
The difference between hw_offload = yes and hw_offload = full is that hw_offload = yes means crypto offload meaning the kernel offloads encryption and decryption to the HW, while hw_offload = full means full HW offload, which offloads encryption, decryption, encapsulation and other operations with header to the HW.

The reason why we need a new option (hw_offload = full) and not using hw_offload = yes because I think we should keep backward compatibility, meaning also if the user has hardware that supports full HW offload, maybe he doesn't want full HW offload and just crypto offload. 
hw_offload = auto should take care of the situation that if the HW and kernel support full HW offload then it should configure full HW offload.

We (Nvidia/Mellanox) already have a fork of strongSwan 5.9.6 (https://github.com/Mellanox/strongswan/tree/5.9.6bf) that adds this support and we provide it with the bfb image (image for our smartNIC/DPU), and we are interested in adding this support to the official strongSwan, so that anyone who uses strongSwan with HW that supports full HW offload can use it.

I would like to hear your opinion on this and if you think we can upstream this change.

**Comment by tobiasbrunner** [comment]
Adding a separate `full` option makes sense. But I think `yes` will be confusing once that's added. It should probably be renamed to `crypto` or similar to clarify what it actually enables (for backwards compatibility we could keep `yes` as undocumented alias). And if `auto` does `full` -> `crypto` -> `nothing` depending on what the hardware supports that's fine too.
